RFCs/0016-Loggers-Information-From-RunSettings.md

```diff
@@ -0,0 +1,98 @@
+
+## Specification
+1. If same logger is specified both in command line arguments and run settings, command line takes precedence.
+2. User can override existing logger value in runsettings from command line. For example `vstest.console foo.dll -- LoggerRunSettings.Loggers="<Logger friendlyName=\"sampleLoggerwithParameters\" />"`.
+2. Multiple Loggers can be added in runsettings by adding `Logger` node in `LoggerRunSettings.Loggers` section.
+3. Configuration is optional in `Logger` node.
+4. Atleast one attribute among `uri`, `friendlyName`, `assemblyQualifiedName` should be present in `Logger` node.
+5. If more than one attributes among `uri`, `friendlyName`, `assemblyQualifiedName` are present, then precedence order is assemblyQualifiedName > uri > friendlyName. Attributes other than precedent attribute are ignored.
+6. Logger can be enabled or disabled using `enabled` attribute in `Logger` node. For example to disable a logger: `<Logger friendlyName="sampleLogger" enabled="false" />`.
+
+## Error and Warning scenarios:
+Exception scenarios: 
+In case of exception, test run is aborted. Following are the exception scenarios:
+1. `LoggerRunSettings` or `Loggers` node has attributes.
+2. `LoggerRunSettings` has any node other than `Loggers`.
+3. `Loggers` has any node other than `Logger`.
+4. `Logger` has any node other than `Configuration`.
+5. `Logger` node has any attribute other than `uri`, `assemblyQualifiedName`, `friendlyName`, or `enabled`.
+6. Invalid format `uri` is given.
+7. Unable to find logger using precedent attribute.
+8. If none of the attributes `uri`, `assemblyName`, `friendlyName` is given.
+
+Warning scenarios:
+1. Key value pair has empty or whitespace value. Example: `<Key1></Key1>`.
+
+Ignored scenarios:
+1. Duplicate keys in `Configuration` node is ignored and verbose is added for it.
+
+## Design
+Following are the proposed design changes for enabling logger support for protocol and C# library based clients:
+1. TestLoggerManager will manage all the loggers. TestLoggerManager will be responsible for:
+  * Initializing loggers from the run settings.
+  * Enabling and triggering logger events.
+  * Disposing logger events.
+
+2. TestPlatform will create and initialize TestLoggerManager and will pass this TestLoggerManager instance to TestRunRequest and DiscoveryRequest.
+3. TestRunRequest and DiscoveryRequest will hold the TestLoggerManager instance. On receving any run/discovery events, respective methods of TestLoggerManager will be invoked. i.e. TestLoggerManager will no longer register logger events with run or discovery events.
+4. On test run/discovery completion, TestLoggerManager will be disposed by TestRunRequest and DiscoveryRequest.
+5. Limitaion with this approach is that if TestRunRequestusers like TestRequestManager tries to call `TestRunRequest.ExecuteAsync()` second time, then logger events will not be invoked as TestLoggerManager will already be disposed in first run.
```

Comment:
s/Limitaion/Limitation

Reply:
s/TestRunRequestusers

---
In reply to: [160961891](https://github.com/Microsoft/vstest-docs/pull/111#discussion_r160961891) [](ancestors = 160961891)

Reply:
This is interesting. What is the plan if we need this in future ?

---
In reply to: [160987392](https://github.com/Microsoft/vstest-docs/pull/111#discussion_r160987392) [](ancestors = 160987392,160961891)

Reply:
For now, we are supporting loggers for first test run of a test request. I will add a comment in code as well. It can come as a separate feature work if we add second run scenario.